### PR TITLE
[server][metrics] Fix sawtooth behavior on follower lag monitoring

### PR DESCRIPTION
**Pull request by ZacAttack**
## [server][metrics] Fix sawtooth behavior on follower lag monitoring

This tweaks the behavior of the follower metric to retain the highest value of the last received timestamp.  Prior to this, it was strictly reporting the last received timestamp.  The result of this was that when one data center was lagged, the follower metric would sawtooth between healthy and unhealthy signals.  This tweak makes it so the most recent timestamp will be retained and reported for the metric for all received data.

## Does this PR introduce any user-facing changes?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure to explain your proposed changes and call out the behavior change.

### CODE REVIEW

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/stats/ingestion/heartbeat/HeartbeatTimeStampEntry.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
+package com.linkedin.davinci.stats.ingestion.heartbeat;
+
+public class HeartbeatTimeStampEntry {
```
Comment (sixpluszero) [inline clients/da-vinci-client/src/main/java/com/linkedin/davinci/stats/ingestion/heartbeat/HeartbeatTimeStampEntry.java:3]: Nice, we get rid of Pair

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/stats/ingestion/heartbeat/HeartbeatMonitoringService.java [hunk 2]
```diff
@@ -274,13 +273,24 @@ private void recordHeartbeat(
...
       heartbeatTimestamps.computeIfPresent(store, (storeKey, perVersionMap) -> {
         perVersionMap.computeIfPresent(version, (versionKey, perPartitionMap) -> {
           perPartitionMap.computeIfPresent(partition, (partitionKey, perRegionMap) -> {
-            perRegionMap.put(region, new MutablePair<>(timestamp, isReadyToServe));
+            // If we are retaining only the highest timestamp for a given heartbeat, if the current held heartbeat
+            // is of a higher value AND was an entry was consumed (not a place holder value by the process) then
+            // we will No-Op in favor of retaining that higher timestamp. This behavior is specific to follower
+            // nodes because the intent of this metric is to only show the lag of the follower relative to the leader
+            if (retainHighestTimeStamp && perRegionMap.get(region) != null
+                && perRegionMap.get(region).timestamp > timestamp && !perRegionMap.get(region).consumedFromUpstream) {
```
Comment (sixpluszero) [inline clients/da-vinci-client/src/main/java/com/linkedin/davinci/stats/ingestion/heartbeat/HeartbeatMonitoringService.java:288]: !perRegionMap.get(region).consumedFromUpstream
Should this be true?

**Review by sixpluszero — COMMENTED** [review]
Thank you for putting up the PR and get rid of the pair.
I left one question about the skip logic.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/stats/ingestion/heartbeat/HeartbeatMonitoringService.java (hunk 2 above)
Comment (ZacAttack) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/stats/ingestion/heartbeat/HeartbeatMonitoringService.java:288]: oh that's right, it should be true, good catch.

**Review by sixpluszero — APPROVED** [review]
Thanks for the fix!

**Pull request merged**